core/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java

```diff
@@ -89,6 +91,12 @@
     private final boolean ignoreUnknownFields;
 
     /**
+     * A special purpose field parser that gets used when the provided matchFieldPrecidate accepts it
+     */
+    private FieldParser matchFieldParser;
+    private Predicate<String> matchFieldPredicate;
```

Comment:
I am trying to think if these can be a single field rather than two that are so tightly coupled. I couldn't come up with a good idea but maybe you will...